x/gamm/pool-models/stableswap/pool.go

```diff
@@ -286,19 +286,43 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	return pCopy.joinPoolSharesInternal(ctx, tokensIn, swapFee)
 }
 
-// TODO: implement this
-func (p *Pool) CalcJoinPoolNoSwapShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error) {
-	return sdk.ZeroInt(), nil, err
+// CalcJoinPoolNoSwapShares creates and returns the number of shares
+// from joining and updating the pool with the given tokensIn and swapFee.
+// Only works more than one token in tokensIn and equal proportions
+// relative to existing liquidity of thereof. Does not mutate pool state.
+// implements Pool interface. See its definition for more details.
+func (p *Pool) CalcJoinPoolNoSwapShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, tokensJoined sdk.Coins, err error) {
+	// N.B. we simulate the calculation by attempting to join a copy of the original pool.
+	// The original pool is not modified.
+	pCopy := p.Copy()
+	numShares, tokensJoined, err = pCopy.joinPoolNoSwapSharesInternal(ctx, tokensIn, swapFee)
```

Comment:
This isn't right, because JP no swap attempts to swap when all assets are given in non-perfect ratios.

Should use `cfmm_common.MaximalExactRatioJoin(p, ctx, tokensIn)`

Reply:
This is `joinPoolNoSwapSharesInternal` though that doesn't do that. I implemented a separate method, similar to how we do this in balancer

Reply:
https://github.com/osmosis-labs/osmosis/blob/99dcf1c9fe07bfde8b8524cb187305a564269423/x/gamm/pool-models/stableswap/amm.go#L328